noman.el

```diff
@@ -69,141 +70,115 @@
-  (let ((buttons (list))
-	(button-func (noman--get-button-func cmd)))
-    (with-current-buffer buffer
+  (alist-get (car (split-string cmd " " 'omit-nulls))
+             noman-parsing-functions #'noman--button nil #'string=))
+
+(defun noman--buttonize (cmd)
+  "Evaluate CMD's button function on each line in `current-buffer'.
+Return list of created buttons."
+  (let ((button-func (noman--button-func cmd))
+        (buttons nil))
+    (save-excursion
       (goto-char (point-min))
-      (let ((max-lines (count-lines (point-min) (point-max))))
-	(while (< (line-number-at-pos (point)) (+ max-lines 1))
-	  (when-let ((current-line-string
-		      (buffer-substring-no-properties
-		       (line-beginning-position) (line-end-position)))
-		     (button (apply button-func (list current-line-string))))
-	    (push button buttons))
-	  (forward-line))))
-    buttons))
+      (while (not (eobp))
+        (push (funcall button-func
+                       (buffer-substring (line-beginning-position) (line-end-position)))
+              buttons)
+        (forward-line)))
+    (delq nil buttons)))
 
 (defvar-keymap noman-mode-map
   "m" #'noman-menu
   "g" #'noman-menu
   "G" #'noman
-  "q" #'quit-window
   "n" #'next-line
   "p" #'previous-line
   "l" #'noman-back
   "TAB" #'forward-button
   "<backtab>" #'backward-button)
 
 (define-derived-mode noman-mode special-mode "noman"
-  "A mode for browsing command line help.")
+  "Major mode for browsing command line help.
 
-(defun noman (cmd)
-  "Attempt to parse command line help for the command CMD.
-
-Noman (no-man) has similar keybindings to man:
-
-g/m  -  jump to a subcommand
-G    -  view help for a different command
-l    -  go back to the last subcommand"
-  (interactive (let* ((paths (exec-path))
-                      (commands '()))
-                 (dolist (path paths)
-                   (when (file-directory-p path)
-                     (setq commands
-                           (append commands
-                                   (directory-files path t "^[^.].*")))))
-                 (setq commands (seq-filter #'file-executable-p commands))
-                 (setq commands (mapcar #'file-name-nondirectory commands))
-                 (list
-                  (completing-read "Program: " commands nil t))))
-  (push cmd noman--history)
-  (let* ((buffer (get-buffer-create (format "*noman %s*" cmd)))
-         (cmdprefix (car (split-string cmd)))
-         (cmdtype (string-trim (shell-command-to-string
-                                (concat "command -V " cmdprefix))))
+\\{noman-mode-map}")
+
+(defun noman--buffer (cmd)
+  "Prepare and display noman CMD's noman buffer."
+  (let* ((tokens (split-string cmd))
+         (prefix (car tokens))
+         (type (string-trim (shell-command-to-string (concat "command -V " prefix))))
          (inhibit-read-only t))
-    (with-current-buffer buffer
+    (with-current-buffer (get-buffer-create "*noman*")
```

Comment:
Instead of creating a new buffer for every command, we can re-use the existing buffer.
Users who wish to keep multiple buffers around still have the option of renaming the buffers.